src/Features/CSharp/Portable/RemoveUnusedParametersAndValues/CSharpRemoveUnusedValuesCodeFixProvider.cs

```diff
@@ -0,0 +1,235 @@
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.RemoveUnusedValues), Shared]
+    internal class CSharpRemoveUnusedValuesCodeFixProvider:
+        AbstractRemoveUnusedValuesCodeFixProvider<ExpressionSyntax, StatementSyntax, BlockSyntax,
+            ExpressionStatementSyntax, LocalDeclarationStatementSyntax, VariableDeclaratorSyntax,
+            ForEachStatementSyntax, SwitchSectionSyntax, SwitchLabelSyntax, CatchClauseSyntax, CatchClauseSyntax>
+    {
+        protected override BlockSyntax WrapWithBlockIfNecessary(IEnumerable<StatementSyntax> statements)
+            => SyntaxFactory.Block(statements);
+
+        protected override SyntaxToken GetForEachStatementIdentifier(ForEachStatementSyntax node)
+            => node.Identifier;
+
+        protected override SyntaxNode TryUpdateNameForFlaggedNode(SyntaxNode node, SyntaxToken newName)
+        {
+            switch (node.Kind())
+            {
+                case SyntaxKind.IdentifierName:
+                    var identifierName = (IdentifierNameSyntax)node;
+                    return identifierName.WithIdentifier(newName.WithTriviaFrom(identifierName.Identifier));
+
+                case SyntaxKind.VariableDeclarator:
+                    var variableDeclarator = (VariableDeclaratorSyntax)node;
+                    return variableDeclarator.WithIdentifier(newName.WithTriviaFrom(variableDeclarator.Identifier));
+
+                case SyntaxKind.SingleVariableDesignation:
+                    return SyntaxFactory.SingleVariableDesignation(newName).WithTriviaFrom(node);
+
+                case SyntaxKind.CatchDeclaration:
+                    var catchDeclaration = (CatchDeclarationSyntax)node;
+                    return catchDeclaration.WithIdentifier(newName.WithTriviaFrom(catchDeclaration.Identifier));
+
+                default:
+                    Debug.Fail($"Unexpected node kind for local/parameter declaration or reference: '{node.Kind()}'");
+                    return null;
+            }
+        }
+
+        protected override void InsertAtStartOfSwitchCaseBlockForDeclarationInCaseLabelOrClause(SwitchSectionSyntax switchCaseBlock, SyntaxEditor editor, LocalDeclarationStatementSyntax declarationStatement)
+        {
+            var firstStatement = switchCaseBlock.Statements.FirstOrDefault();
+            if (firstStatement != null)
+            {
+                editor.InsertBefore(firstStatement, declarationStatement);
+            }
+            else
+            {
+                // Switch section without any statements is an error case.
+                // Insert before containing switch statement.
+                editor.InsertBefore(switchCaseBlock.Parent, declarationStatement);
+            }
+        }
+
+        protected override Task<SyntaxNode> RemoveDiscardDeclarationsAsync(
+            SyntaxNode memberDeclaration,
+            Document document,
+            CancellationToken cancellationToken)
```

Comment:
dump question, and i'm not pushing back heavily on this.  But this seems like very different functionality.  Would it potentially make more sense to have a different fixer for dealing with this sort of thing?

Reply:
Without this functionality, the code fix that adds discard starts causing compile errors. It would make sense to move this functionality to a service instead of doing it within the fixer itself. Let me refactor this out.

---
In reply to: [234799148](https://github.com/dotnet/roslyn/pull/30777#discussion_r234799148) [](ancestors = 234799148)

Reply:
that makes sense!  Don't put yourself out here though if it's a lot of effort.

Reply:
Done.

---
In reply to: [234809009](https://github.com/dotnet/roslyn/pull/30777#discussion_r234809009) [](ancestors = 234809009,234799148)

Reply:
That is a very good suggestion, and I have moved it to a separate service, which is currently C# specific. Heejae gave me couple other suggestions on the same line and I agree with both of you that is the correct direction for the IDE features. I myself benefitted here from IMoveDeclarationNearReference functionality being so modular and extracted into a service.

---
In reply to: [234809375](https://github.com/dotnet/roslyn/pull/30777#discussion_r234809375) [](ancestors = 234809375)